### ISSUE TITLE
Processing out of order blockwise responses and retry the request 

### ISSUE THREAD

**Issue by boaks** [eclipse/californium#275]
hudson:
https://hudson.eclipse.org/californium/job/2.0.x/146/

The build failed:
Simplified sequence:
- GET BLOCK 0
- GET BLOCK 0 (Retry)
- GET BLOCK 0 (Retry)
- ACK BLOCK 0
- GET BLOCK 1 (Dropped by test intention!)
- ACK BLOCK 0 (Retry)
- ACK BLOCK 0 (Retry)

My idea:
The `ACK BLOCK 0 (Retry)` stops the retransmission of the Exchange in the ReliabilityLayer.receiveResponse.
The BlockwiseLayer ignores the wrong block.

> 249 FINE [DirectDatagramSocketImpl]: (E199,T249) 127.0.0.1:8,234 ==> :8,236 [CON-GET    MID=  332, Token=ffd06f7a0af2b6a7, OptionSet={"Uri-Host":"localhost", "Uri-Path":"test"}, no payload] - (org.eclipse.californium.elements.util.DirectDatagramSocketImpl.java:817) receive() in thread Thread-1 at (03:54:33.021)
> 249 FINE [DirectDatagramSocketImpl]: (E201,T249) 127.0.0.1:8,234 ==> :8,236 [CON-GET    MID=  332, Token=ffd06f7a0af2b6a7, OptionSet={"Uri-Host":"localhost", "Uri-Path":"test"}, no payload] - (org.eclipse.californium.elements.util.DirectDatagramSocketImpl.java:817) receive() in thread Thread-1 at (03:54:33.628)
> 249 FINE [DirectDatagramSocketImpl]: (E202,T249) 127.0.0.1:8,234 ==> :8,236 [CON-GET    MID=  332, Token=ffd06f7a0af2b6a7, OptionSet={"Uri-Host":"localhost", "Uri-Path":"test"}, no payload] - (org.eclipse.californium.elements.util.DirectDatagramSocketImpl.java:817) receive() in thread Thread-1 at (03:54:33.628)
> 249 FINE [DirectDatagramSocketImpl]: (E203,T249) 127.0.0.1:8,235 ==> :8,236 [ACK-2.05   MID=  332, Token=ffd06f7a0af2b6a7, OptionSet={"Content-Format":"text/plain", "Block2":"(szx=1/32, m=true, num=0)", "Size2":250}, "693124610902415439268315".. 32 bytes] - (org.eclipse.californium.elements.util.DirectDatagramSocketImpl.java:817) receive() in thread Thread-1 at (03:54:33.630)
> 249 FINE [DirectDatagramSocketImpl]: increasing receive buffer from 27 to full buffer capacity [2,000] - (org.eclipse.californium.elements.util.DirectDatagramSocketImpl.java:817) receive() in thread Thread-1 at (03:54:33.631)
> 249 FINE [DirectDatagramSocketImpl]: (E206,T249) 127.0.0.1:8,235 ==> :8,236 [ACK-2.05   MID=  332, Token=ffd06f7a0af2b6a7, OptionSet={"Content-Format":"text/plain", "Block2":"(szx=1/32, m=true, num=0)", "Size2":250}, "693124610902415439268315".. 32 bytes] - (org.eclipse.californium.elements.util.DirectDatagramSocketImpl.java:817) receive() in thread Thread-1 at (03:54:33.632)
> 249 FINE [DirectDatagramSocketImpl]: (E207,T249) 127.0.0.1:8,235 ==> :8,236 [ACK-2.05   MID=  332, Token=ffd06f7a0af2b6a7, OptionSet={"Content-Format":"text/plain", "Block2":"(szx=1/32, m=true, num=0)", "Size2":250}, "693124610902415439268315".. 32 bytes] - (org.eclipse.californium.elements.util.DirectDatagramSocketImpl.java:817) receive() in thread Thread-1 at (03:54:33.633)
> Dropping packet 5 (burst 1) from server
> 244 WARNING [UdpMatcher]: Possible MID reuse before lifetime end for token [ffd06f7a0af2b6a7], expected MID 333 but received 332 - (org.eclipse.californium.core.network.UdpMatcher.java:827) receiveResponse() in thread CoapEndpoint-coap://localhost:0#1 at (03:54:33.633)
> 249 FINE [DirectDatagramSocketImpl]: (E210,T249) 127.0.0.1:8,234 ==> :8,236 [CON-GET    MID=  333, Token=ffd06f7a0af2b6a7, OptionSet={"Uri-Host":"localhost", "Uri-Path":"test", "Block2":"(szx=1/32, m=false, num=1)"}, no payload] - (org.eclipse.californium.elements.util.DirectDatagramSocketImpl.java:817) receive() in thread Thread-1 at (03:54:33.634)
> Dropping packet 6 (burst 2) from client
> 244 WARNING [BlockwiseLayer]: ignoring block2 response with wrong block number 0 (expected 1): ACK-2.05   MID=  332, Token=ffd06f7a0af2b6a7, OptionSet={"Content-Format":"text/plain", "Block2":"(szx=1/32, m=true, num=0)", "Size2":250}, "693124610902415439268315".. 32 bytes - (org.eclipse.californium.core.network.stack.BlockwiseLayer.java:564) handleBlock2Response() in thread CoapEndpoint-coap://localhost:0#1 at (03:54:33.634)
>

**Comment by boaks** [comment]
I will adjust the test to reproduce the sequence.

**Comment by boaks** [comment]
Proposal for fixing it:
Add `stopRetransmission` and `restartRetransmission` to `Exchange`, and restart it on "block2 ignore".
I would prepare a PR.
Other fixing proposals are welcome :-)

**Comment by boaks** [comment]
`Exchange` doesn't work, I go to add `restartRetransmission` to `Layer`.
Uuggly, workaround, so better ideas are really welcome!

**Comment by sbernard31** [comment]
I think you're right : The ACK BLOCK 0 (Retry) stops the retransmission of the Exchange in the ReliabilityLayer.receiveResponse.

When I launched your test, I get this log :
`AVERTISSEMENT: Possible MID reuse before lifetime end for token [f050e9aacbe505fc], expected MID 276 but received 275`

The UdpMatcher code corresponding is  :
```java
// we have received a Response matching the token of an ongoing Exchange's Request
// according to the CoAP spec (https://tools.ietf.org/html/rfc7252#section-4.5),
// message deduplication is relevant for CON and NON messages only

if ((response.getType() == Type.CON || response.getType() == Type.NON) &&
		exchangeStore.findPrevious(idByMID, exchange) != null) {
	LOGGER.log(Level.FINER, "Received duplicate response for open exchange: {0}", response);
	response.setDuplicate(true);
} else if (!isNotify) {
	// we have received the expected response for the original request
	idByMID = KeyMID.fromOutboundMessage(exchange.getCurrentRequest());
	if (exchangeStore.remove(idByMID, exchange) != null) {
		LOGGER.log(Level.FINE, "Closed open request [{0}]", idByMID);
	}
}

if (response.getType() == Type.ACK && exchange.getCurrentRequest().getMID() != response.getMID()) {
	// The token matches but not the MID.
	LOGGER.log(Level.WARNING,
			"Possible MID reuse before lifetime end for token [{0}], expected MID {1} but received {2}",
			new Object[]{response.getTokenString(), exchange.getCurrentRequest().getMID(), response.getMID()});
}
return exchange;
```

I do not understand why we just return exchange even if the MID does not match ... If I return null when MID mismatch the test is ok.
I also don't understand why the code which should detect duplicate don't consider the second ACK as duplicate :/

**Comment by boaks** [comment]
> I do not understand why we just return exchange even if the MID does not match 

Because, if a response (matching token) is received, this matches also. The issue is in my opinion caused by reusing the token for several request.

> I also don't understand why the code which should detect duplicate don't consider the second ACK as duplicate :/

Very good hint! I will analyse that, maybe that's a better fix then the workaround.

**Comment by boaks** [comment]
I checked it, we only deduplicate CON/NON. Because RCF7252 4.5 mentions only CON/NON.
But my interpretation of this section in the RFC would be, that the intention there was the upper-layer processing, not the CoAP implementation cleanup ´, which in my opinion could also be skipped.

I will try it.

**Comment by sbernard31** [comment]
>Because, if a response (matching token) is received, this matches also.

When we receive a PIGGYBACK, we first search by token, we found the exchange, but if the MID does not match, why just not ignoring it ? like for empty ACK : 
```java
public Exchange receiveEmptyMessage(final EmptyMessage message) {

	// an empty ACK or RST always is received as a reply to a message
	// exchange originating locally, i.e. the message will echo an MID
	// that has been created here
	KeyMID idByMID = KeyMID.fromInboundMessage(message);
	Exchange exchange = exchangeStore.remove(idByMID, null);

	if (exchange != null) {
		LOGGER.log(Level.FINE, "Received expected reply for message exchange {0}", idByMID);
	} else {
		LOGGER.log(Level.FINE,
				"Ignoring unmatchable empty message from {0}:{1}: {2}",
				new Object[]{message.getSource(), message.getSourcePort(), message});
	}
	return exchange;
}
```

> The issue is in my opinion caused by reusing the token for several request.

I think its a bad idea too.

**Comment by boaks** [comment]
So extending the deduplication to ACKs would also solve this issue.
The only drawback of this would be, that if the MID is reused before the end of the MID lifetime, the response will be lost due to deduplication.
But reusing the MID before its lifetime is very dangerous, so I would life with that drawback.
Therefore I changed the failing Deduplication test also.

**Comment by sbernard31** [comment]
Sry to insist but why not just : returning null if MID does not match for ACK ?

**Comment by boaks** [comment]
See #281

**Comment by boaks** [comment]
Fixed with #281
